### ISSUE TITLE
Internal leakage occurs when the broker is established and the client is used to listen

### ISSUE THREAD

**Issue by git123java** [bytebeamio/rumqtt#732]
` loop{
                let event_result = eventloop.poll().await;
                match event_result {} 
          }`
 After I create a broker, I create a client and connect to the broker, and then another client connects to the broker. The first client can listen to the messages published by client2, but as the number of messages published by client2 increases, The memory used by the broker program gradually increased, not decreased, until I discovered that it was the code used in the flush method method: `self.socket.write_all(&write[...] ).await? ` 
 it because client2 keeps Posting messages and sockets keep growing without emptying? How should this be solved?

**Comment by swanandx** [comment]
When broker gets a message published by client, it stores the message in a commitlog, which i believe, is the reason you are seeing the increase in memory usage as the number message published increase!

Growth in memory usage as the number of messages increase is expected till a certain limit. You can control the maximum size of the segment in commitlog by `max_segment_size` in `rumqttd.toml`.

**Comment by git123java** [comment]
I look at these parameters, set the maximum number, the new package will not be accepted, is rejected; So how do you clear the old data to make room for the new data

**Comment by swanandx** [comment]
if there is no space, older segments are dropped automatically by the broker! can you provide rumqttd config and the client code you are using to we can debug this further?

**Comment by git123java** [comment]
[router]
id = 0
max_connections = 10010
max_outgoing_packet_count = 200
max_segment_size = 104857600
max_segment_count = 10

**Comment by git123java** [comment]
**[router]
id = 0
max_connections = 10010
max_outgoing_packet_count = 200
max_segment_size = 104857600
max_segment_count = 10
[v5.1]
name = "v5-1"
listen = "0.0.0.0:18804"
next_connection_delay_ms = 1
    [v5.1.connections]
    connection_timeout_ms = 60000
    max_payload_size = 20480
    max_inflight_count = 100**

This is my toml file above，the client code is as follows:
`#[tokio::main]
async fn main() {
    let commandline: CommandLine = CommandLine::parse();

    if let Some(Command::GenerateConfig) = commandline.command {
        println!("{RUMQTTD_DEFAULT_CONFIG}");
        return;
    }

    // if !commandline.quiet {
    //     banner();
    // }

    let level = match commandline.verbose {
        0 => "rumqttd=warn",
        1 => "rumqttd=info",
        2 => "rumqttd=debug",
        _ => "rumqttd=trace",
    };
    // tracing syntax ->
    let builder = tracing_subscriber::fmt()
        .pretty()
        .with_line_number(false)
        .with_file(false)
        .with_thread_ids(false)
        .with_thread_names(false)
        .with_env_filter(level)
        .with_filter_reloading();

    let reload_handle = builder.reload_handle();

    builder
        .try_init()
        .expect("initialized subscriber succesfully");


    let mut config_builder = config::Config::builder();

    config_builder = match &commandline.config {
        Some(config) => config_builder.add_source(config::File::with_name(config)),
        None => config_builder.add_source(config::File::from_str(
            RUMQTTD_DEFAULT_CONFIG,
            FileFormat::Toml,
        )),
    };

    let mut configs: rumqttd::Config = config_builder.build().unwrap().try_deserialize().unwrap();
    configs.console.set_filter_reload_handle(reload_handle);

    validate_config(&configs);

    // println!("{:#?}", configs);
    
    thread::spawn(||{
        let mut broker = Broker::new(configs);
        broker.start().unwrap();
        println!("start");
    });
    
    
    thread::sleep(Duration::from_secs(1));

    let mut mqttoptions = MqttOptions::new("test", "127.0.0.1", 18803);
       
        mqttoptions
            .set_keep_alive(Duration::from_secs(5));
        mqttoptions.set_clean_session(true);
        let (client, mut eventloop) = AsyncClient::new(mqttoptions,5 );
        task::block_on(client.subscribe("upload/data", QoS::AtLeastOnce)).unwrap();

            // thread::spawn(move||
                loop{
                let event_result = eventloop.poll().await;
                match event_result {
                    Ok(event)=>{
                        match event{
                            Event::Incoming(incoming)=> {
                                if let Incoming::Publish(publish) = incoming.clone() {
                                    println!("message {:?}", publish.topic);
                                    let res=task::block_on(handel_topic(publish, "147".to_string(), client.clone()));
                                }
                            }
                            Event::Outgoing(outgoing)=>{
                                println!("outgoing {:?}",outgoing)
                            }
                        }
                    }
                    Err(e)=>{
                        println!("err {:?}",e);
                    }
                }
            }
        // );
}
`

**Comment by swanandx** [comment]
BTW, it would be also helpful if you paste the failure logs, like wdym by package being rejected and what did rumqttd log the reason of rejecting the packet.

**Comment by git123java** [comment]
This is my toml file
`[router]
id = 0
max_connections = 10010
max_outgoing_packet_count = 200
max_segment_size = 104857600
max_segment_count = 10

[v5.1]
name = "v5-1"
listen = "0.0.0.0:18804"
next_connection_delay_ms = 1
    [v5.1.connections]
    connection_timeout_ms = 60000
    max_payload_size = 20480
    max_inflight_count = 100
`
Here is my broker and client code：
`
thread::spawn(||{
        let mut broker = Broker::new(configs);
        broker.start().unwrap();
        println!("start");
    });
    
    
    thread::sleep(Duration::from_secs(1));

    let mut mqttoptions = MqttOptions::new("test", "127.0.0.1", 18804);
       
        mqttoptions
            .set_keep_alive(Duration::from_secs(5));
       
        let (client, mut eventloop) = AsyncClient::new(mqttoptions,5 );
        task::block_on(client.subscribe("/upload/data", QoS::AtLeastOnce)).unwrap();
            // thread::spawn(move||
                loop{
                let event_result = task::block_on(eventloop.poll());
                match event_result {
                    Ok(event)=>{
                        match event{
                            Event::Incoming(incoming)=> {
                                
                                if let Incoming::Publish(publish) = incoming.clone() {
                                    println!("message {:?}", publish.topic);
                                    let res=task::block_on(handel_topic(publish, "test1".to_string(), client.clone()));
                                }
                      
                            }
                            Event::Outgoing(outgoing)=>{
                                println!("outgoing {:?}",outgoing)
                            }
                        }
                    }
                    Err(e)=>{
                        println!("err {:?}",e);
                    }
                }
            }`

**Comment by git123java** [comment]
When the program reaches this step, "let event_result = task::block_on(eventloop.poll()); If the client continues to send messages, the memory will continue to expand

**Comment by git123java** [comment]
It seems to have been successfully solved, the method is that I set the toml file parameter `max_segment_size = 10000,` and then the program memory will not expand, I go to apply this configuration to my program. I'll see if it affects the content and quality of the message

**Comment by swanandx** [comment]
That's good to know! Closing this as it isn't an bug / issue ( as described in the comments ). Feel free to comment if you wish!

Btw, just curious why you are using Async client instead of sync Client if you want to block on everything instead of calling `.await`?

Thanks for reporting 😁

**Comment by git123java** [comment]
Thank you for your reminder to modify toml!
Asynchronous is chosen to avoid unnecessary waiting and blocking on synchronous clients, aren't asynchronous clients more friendly

**Comment by swanandx** [comment]
> Asynchronous is chosen to avoid unnecessary waiting and blocking on synchronous client
>

Yup! But in your code, you are using `task::block_on` which blocks right? Async will be beneficial if u use `.await` instead of block_on right?

e.g. `client.subscribe("topic").await?`

**Comment by git123java** [comment]
Yes, we do! Thank u. Those are two things I can improve on
